internal/timeline/tweener.go

```diff
@@ -0,0 +1,308 @@
+	CUR_TO_END TweenerMode = iota
+	START_TO_END
+	START_TO_CUR
+)
+
+type ITweener interface {
+	ITimeline
+	GetTweener() *Tweener
+	// Rewind()
+}
+
+type Tweener struct {
+	Timeline
+	mode     TweenerMode
+	getter   func() interface{}
+	setter   func(interface{})
+	startVal interface{}
+	curVal   interface{} // only for START_TO_END mode
+	endVal   interface{}
+	duration float64
+}
+
+func (t *Tweener) GetTweener() *Tweener {
+	return t
+}
+
+func (t *Tweener) Init__0(getter func() interface{}, setter func(interface{}), endVal interface{}, duration float64) *Tweener {
+	t.mode = CUR_TO_END
+	t.getter = getter
+	t.setter = setter
+	t.endVal = endVal
+	if duration < 0.0 {
+		duration = 0.0
+	}
+	t.duration = duration
+	t.Timeline.Init()
+	t.Timeline.onActive = t.onActive
+	t.Timeline.onStep = func(time *float64) ITimeline {
+		return t.onStep(time)
+	}
+	if debugTweener {
+		log.Printf("Tweener.Init__0 %X", t.ID)
+	}
+	return t
+}
+
+func (t *Tweener) Init__1(setter func(interface{}), startVal interface{}, endVal interface{}, duration float64) *Tweener {
+	t.mode = START_TO_END
+	t.setter = setter
+	t.startVal = startVal
+	t.curVal = startVal
+	t.endVal = endVal
+	if duration < 0.0 {
+		duration = 0.0
+	}
+	t.duration = duration
+	t.Timeline.Init()
+	t.Timeline.onActive = t.onActive
+	t.Timeline.onStep = func(time *float64) ITimeline {
+		return t.onStep(time)
+	}
+	if debugTweener {
+		log.Printf("Tweener.Init__1 %X", t.ID)
+	}
+	return t
+}
+
+func (t *Tweener) Init__2(getter func() interface{}, setter func(interface{}), startVal interface{}, duration float64) *Tweener {
+	t.mode = START_TO_CUR
+	t.getter = getter
+	t.setter = setter
+	t.startVal = startVal
+	t.duration = duration
+	t.Timeline.Init()
+	t.Timeline.onActive = t.onActive
+	t.Timeline.onStep = func(time *float64) ITimeline {
+		return t.onStep(time)
+	}
+	if debugTweener {
+		log.Printf("Tweener.Init__2 %X", t.ID)
+	}
+	return t
+}
+
+func (t *Tweener) GetTimeline() *Timeline {
+	return &t.Timeline
+}
+
+func (t *Tweener) onActive(active bool) {
+	if debugTweener {
+		log.Printf("Tweener.onActive %t %X", active, t.ID)
+	}
+
+	switch t.mode {
+	case CUR_TO_END:
+		if active {
+			t.startVal = t.getter()
+		} else {
+			t.setter(t.endVal)
+		}
+	case START_TO_CUR:
+		if active {
+			t.endVal = t.getter()
+			t.setter(t.startVal)
+		} else {
+			t.setter(t.endVal)
+		}
+	case START_TO_END:
+		if active {
+			t.setter(t.startVal)
+		} else {
+			t.setter(t.endVal)
+		}
+	}
+}
+
+func (t *Tweener) Step(time *float64) ITimeline {
+	timeline := &t.Timeline
+	running := timeline.Step(time)
+	if running == timeline {
+		return t
+	} else {
+		return running
+	}
+}
+
+func (t *Tweener) onStep(time *float64) ITimeline {
+	switch t.startVal.(type) {
+	case float64:
+		return t.onStepFloat64(time)
+	case math32.Vector2:
+		return t.onStepV2(time)
+	case *math32.Vector2:
+		panic("should pass Vector2, not *Vector2, check types of tweener.getter return and startVal and endVal")
+	}
+	return t
```

Comment:
It's a common mistake to pass math32.NewVector2(...) instead of *math32.NewVector2(...). Passing a non-pointer value can cause a panic, alerting the developer to the issue.

Reply:
Since the type of `startVal` & `endVal` is `interface{}`, the user may pass in any type of value, for example, string, which is not tweenable. Should we notice him about that?

Reply:
It is intended to alert engine developers like me, who are responsible for adding functions such as these
func (p *SpriteImpl) TestTweenV2(x, y float64, secs float64) {
	tn := &timeline.Tweener{}
	tn.Init1(func() interface{} {
		return math32.NewVector2(p.x, p.y) // Note here.   It's a mistake
	}, func(obj interface{}) {
		if v2, ok := obj.(math32.Vector2); ok {
			p.x, p.y = v2.X, v2.Y
		}
	}, *math32.NewVector2(x, y), timeline.float64(secs))
	p.g.timelines.AddTimeline(tn)
}

The Spx developer, who will use the function TestTweenV2, has no chance of passing an incorrect type value when calling TestTweenV2.

Reply:
OK